lib/pbench/agent/tool_data_sink.py

```diff
@@ -381,216 +447,362 @@ def tm_log_capture(self):
+            with mdlog_name.open("r") as fp:
+                mdlog.read_file(fp)
+        except FileNotFoundError:
+            # Ignore if it doesn't exist
             pass
 
-    def _fetch_tms(self):
-        """_fetch_tms - fetch all the Tool Meister data for all recorded tool
-        meisters in the Redis server under the tm-pids key.
+        section = "pbench"
+        mdlog.add_section(section)
+        # Users have a funny way of adding '%' characters to the config
+        # variable, so we have to be sure we handle "%" characters in the
+        # config metadata properly.
+        mdlog.set(section, "config", self.optional_md["config"].replace("%", "%%"))
+        mdlog.set(section, "date", self.optional_md["date"])
+        # Users have a funny way of adding '%' characters to the run
+        # directory, so we have to be sure we handle "%" characters in the
+        # directory name metadata properly.
+        mdlog.set(section, "name", self.benchmark_run_dir.name.replace("%", "%%"))
+        version, seqno, sha1, hostdata = collect_local_info(self.pbench_bin)
+        rpm_version = f"v{version}-{seqno}g{sha1}"
+        mdlog.set(section, "rpm-version", rpm_version)
+        mdlog.set(section, "script", self.optional_md["script"])
+
+        section = "controller"
+        mdlog.add_section(section)
+        mdlog.set(section, "hostname", self.hostname)
+        mdlog.set(section, "hostname-s", hostdata["s"])
+        mdlog.set(section, "hostname-f", hostdata["f"])
+        mdlog.set(section, "hostname-i", hostdata["i"])
+        mdlog.set(section, "hostname-A", hostdata["A"])
+        mdlog.set(section, "hostname-I", hostdata["I"])
+        mdlog.set(section, "ssh_opts", self.optional_md["ssh_opts"])
+
+        section = "run"
+        mdlog.add_section(section)
+        mdlog.set(section, "controller", self.hostname)
+        mdlog.set(section, "start_run", _now("start"))
+
+        section = "tools"
+        mdlog.add_section(section)
+        mdlog.set(section, "hosts", " ".join(sorted(list(self.tools.keys()))))
+        mdlog.set(section, "group", self.tool_group)
+        mdlog.set(section, "trigger", str(self.tool_trigger))
+
+        for host, tm in sorted(tms.items()):
+            section = f"tools/{host}"
+            mdlog.add_section(section)
+            mdlog.set(section, "label", tm["label"])
+            tools_string = ",".join(sorted(list(tm["tools"].keys())))
+            mdlog.set(section, "tools", tools_string)
+
+            # add host data
+            mdlog.set(section, "hostname-s", tm["hostname_s"])
+            mdlog.set(section, "hostname-f", tm["hostname_f"])
+            mdlog.set(section, "hostname-i", tm["hostname_i"])
+            mdlog.set(section, "hostname-A", tm["hostname_A"])
+            mdlog.set(section, "hostname-I", tm["hostname_I"])
+            ver, seq, sha = tm["version"], tm["seqno"], tm["sha1"]
+            mdlog.set(section, "rpm-version", f"v{ver}-{seq}g{sha}")
```

Comment:
Log a warning when the Tool Meister versions don't match.  We should also an entry in the `metadata.log` indicating that explicitly in the `run` section (perhaps `tool_meister_version_mismatch = True`).